File: client/src/logic/Board.ts
import { Coordinate } from "../components/gameplay/DrawBoard";
import { Tile } from "../components/gameplay/Tile";
import { PlainShipData } from "./GameSave";
import { SHIP_TYPES, Ship } from "./Ship";
import { Guess } from "./gameLogic";

export class Board {
  player: string;
  height: number;
  width: number;
  ships: Set<Ship>;
  tiles: Tile[][];

  constructor(player: string, height: number, width: number, ships: Set<Ship>, tiles: Tile[][]) {
    this.player = player;
    this.height = height;
    this.width = width;
    this.ships = ships;
    this.tiles = tiles;
  }

  makeCopy(): Board {
    return new Board(this.player, this.height, this.width, this.ships, this.tiles);
  }

  isValid() {
    if (this.player === "") {
      return false;
    }
    if (this.ships.size !== 5) {
      return false;
    }
    for (const ship of this.ships) {
      if (!this.checkNeighborsEmpty(ship.tiles)) {
        return false;
      }
    }
    return true;
  }

  checkAllBoatsSank(): boolean {
    return [...this.ships].every((ship) => ship.isSank());
  }

  addShip(ship: Ship, startCoordinate: Coordinate, vertical: boolean) {
    if (!this.ships.has(ship) && this.canAddShip(ship, startCoordinate, vertical)) {
      ship.setTiles(this.getTiles(ship.type.length, startCoordinate, vertical));
      this.ships.add(ship);
    }
  }
  getTiles(length: number, startCoordinate: Coordinate, vertical: boolean): Tile[] {
    const tiles: Tile[] = [];
    for (let i = 0; i < length; i++) {
      const tile: Tile | undefined =
        this.tiles[startCoordinate.y + (vertical ? i : 0)]?.[
          startCoordinate.x + (vertical ? 0 : i)
        ];
      if (tile) {
        tiles.push(tile);
      }
    }
    return tiles;
  }

  canAddShip(ship: Ship, startCoordinate: Coordinate, vertical: boolean): boolean {
    const tiles = this.getTiles(ship.type.length, startCoordinate, vertical);
    return (
      tiles.length === ship.type.length &&
      tiles.every((tile) => tile.placedShip === null) &&
      this.checkNeighborsEmpty(tiles)
    );
  }

  checkNeighborsEmpty(tiles: Tile[], range = 1): boolean {
    const neighbors: Set<Tile> = new Set();
    for (const tile of tiles) {
      const yStart = Math.max(0, tile.coordinate.y - range);
      const yLimit = Math.min(this.height - 1, tile.coordinate.y + range);
      for (let y = yStart; y <= yLimit; y++) {
        const xStart = Math.max(0, tile.coordinate.x - range);
        const xLimit = Math.min(this.width - 1, tile.coordinate.x + range);
        for (let x = xStart; x <= xLimit; x++) {
          neighbors.add(this.tiles[y][x]);
        }
      }
    }
    for (const tile of tiles) {
      neighbors.delete(tile);
    }
    return [...neighbors].every((tile) => tile.placedShip === null);
  }

  canGuess(coordinate: Coordinate): boolean {
    return !this.tiles[coordinate.y][coordinate.x].guessed;
  }

  submitGuess(coordinate: Coordinate): boolean {
    const tile = this.tiles[coordinate.y][coordinate.x];
    if (tile.guessed) {
      throw new Error("This tile was already guessed!");
    }
    tile.guessed = true;
    return tile.placedShip !== null;
  }

  processGuess(guess: Guess) {
    const tile = this.tiles[guess.coordinate.y][guess.coordinate.x];
    tile.guessed = true;
    if (guess.hit && tile.placedShip === null) {
      tile.placedShip = new Ship(SHIP_TYPES.UNKNOWN, [tile]);
    }
  }

  processGuessSunk(guess: Guess, shipData: PlainShipData) {
    const tile = this.tiles[guess.coordinate.y][guess.coordinate.x];
    tile.guessed = true;
    const { type, startingCoordinate, vertical } = shipData;
    this.#clearUnknownShip(SHIP_TYPES[type].length, startingCoordinate, vertical);
    this.addShip(new Ship(SHIP_TYPES[type], []), startingCoordinate, vertical);
  }

  #clearUnknownShip(length: number, startCoordinate: Coordinate, vertical: boolean) {
    for (let i = 0; i < length; i++) {
      const tile: Tile | undefined =
        this.tiles[startCoordinate.y + (vertical ? i : 0)]?.[
          startCoordinate.x + (vertical ? 0 : i)
        ];
      if (tile && tile.placedShip?.type === SHIP_TYPES.UNKNOWN) {
        tile.placedShip = null;
      }
    }
  }
}

File: server/src/main/java/battleship/game/board/Board.java
package battleship.game.board;

import battleship.dtos.BoardDTO;
import battleship.dtos.ShipDTO;
import battleship.exceptions.BoardException;
import battleship.exceptions.IllegalActionException;
import battleship.game.ship.Ship;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Board {
    private final int height;
    private final int width;
    private final Set<Ship> ships;
    private final Tile[][] tiles;

    public Board(int height, int width, Set<Ship> ships, Tile[][] tiles) {
        this.height = height;
        this.width = width;
        this.ships = ships;
        this.tiles = tiles;
    }

    public void addShip(Ship ship, Coordinate startingCoordinate, boolean vertical) throws BoardException {
        if (!ships.contains(ship) && canAddShip(startingCoordinate, vertical, ship.getLength())) {
            ships.add(ship);
            ship.setOccupiedTiles(getTiles(startingCoordinate, vertical, ship.getLength()));
        }
    }

    public BoardDTO getBoardDataFull(String playerName) {
        return new BoardDTO(playerName, height, width, ships.stream()
                .map(ShipDTO::new)
                .toList()
        );
    }

    public BoardDTO getBoardDataRevealed(String playerName) {
        return new BoardDTO(playerName, height, width, ships.stream()
                .filter(Ship::isSank)
                .map(ShipDTO::new)
                .toList()
        );
    }

    private boolean canAddShip(Coordinate startingCoordinate, boolean vertical, int length) throws BoardException {
        Tile[] tiles = getTiles(startingCoordinate, vertical, length);
        return Arrays.stream(tiles).allMatch(Tile::isEmpty) && checkTileNeighborsEmpty(tiles, 1);
    }

    private boolean checkTileNeighborsEmpty(Tile[] tiles, int range) {
        Set<Tile> tileSet = new HashSet<>(List.of(tiles));
        for (Tile tile : tiles) {
            int yStart = Math.max(0, tile.getCoordinate().y() - range);
            int yLimit = Math.min(height - 1, tile.getCoordinate().y() + range);
            for (int y = yStart; y < yLimit; y++) {
                int xStart = Math.max(0, tile.getCoordinate().x() - range);
                int xLimit = Math.min(width - 1, tile.getCoordinate().x() + range);
                for (int x = xStart; x < xLimit; x++) {
                    if (!tileSet.contains(tile) && tile.isEmpty()) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    private Tile[] getTiles(Coordinate startingCoordinate, boolean vertical, int length) throws BoardException {
        Tile[] tiles = new Tile[length];
        if (vertical) {
            int lowerEnd = startingCoordinate.y() + length - 1;
            if (lowerEnd < height) {
                for (int i = 0; i < length; i++) {
                    tiles[i] = this.tiles[startingCoordinate.y() + i][startingCoordinate.x()];
                }
            } else {
                throw new BoardException("Out of bounds at y by %d".formatted(lowerEnd - height - 1));
            }
        } else {
            int rightEnd = startingCoordinate.x() + length - 1;
            if (rightEnd < width) {
                for (int i = 0; i < length; i++) {
                    tiles[i] = this.tiles[startingCoordinate.y()][startingCoordinate.x() + i];
                }
            } else {
                throw new BoardException("Out of bounds at x by %d".formatted(rightEnd - width - 1));
            }
        }
        return tiles;
    }

    public Ship submitGuess(Coordinate coordinate) throws IllegalActionException {
        return tiles[coordinate.y()][coordinate.x()].guess();
    }

    public Ship getShip(Coordinate coordinate) {
        Tile tile = tiles[coordinate.y()][coordinate.x()];
        if (tile.isEmpty()) {
            throw new RuntimeException("No ship on coordinate.");
        }
        return tile.getShip();
    }

    public boolean allShipSank() {
        return ships.stream().allMatch(Ship::isSank);
    }
}
